### PR TITLE
xdg-utils: fix mailto-vulnerability

### PR DESCRIPTION
**Pull request by Mic92**
###### Motivation for this change

https://www.nds.ruhr-uni-bochum.de/media/nds/veroeffentlichungen/2020/08/15/mailto-paper.pdf

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/X11/xdg-utils/default.nix [hunk 1]
```diff
@@ -32,6 +32,10 @@ stdenv.mkDerivation rec {
   # just needed when built from git
   buildInputs = [ libxslt docbook_xml_dtd_412 docbook_xsl xmlto w3m ];
 
+  patches = [
+    ./0001-xdg-email-remove-attachment-handling-from-mailto.patch
```
Comment (Mic92) [inline pkgs/tools/X11/xdg-utils/default.nix:36]: I will also upstream this. Don't have an account on freedesktop yet.

File: pkgs/tools/X11/xdg-utils/0001-xdg-email-remove-attachment-handling-from-mailto.patch [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+diff --git a/scripts/xdg-email.in b/scripts/xdg-email.in
+index 6db58ad..8096f7d 100644
+--- a/scripts/xdg-email.in
++++ b/scripts/xdg-email.in
+@@ -48,7 +48,6 @@ run_thunderbird()
+     BCC=$(/bin/echo -e $(echo "$MAILTO" | grep '^bcc=' | sed 's/^bcc=//;s/%\(..\)/\\x\1/g' | awk '{ printf "%s,",$0 }'))
+     SUBJECT=$(echo "$MAILTO" | grep '^subject=' | tail -n 1)
+     BODY=$(echo "$MAILTO" | grep '^body=' | tail -n 1)
+-    ATTACH=$(/bin/echo -e $(echo "$MAILTO" | grep '^attach=' | sed 's/^attach=//;s/%\(..\)/\\x\1/g' | awk '{ printf "%s,",$0 }' | sed 's/,$//'))
```
Comment (Mic92) [inline pkgs/tools/X11/xdg-utils/0001-xdg-email-remove-attachment-handling-from-mailto.patch:29]: I think this is still incomplete. It also tries to attach in emacs.

Comment (Mic92) [reply pkgs/tools/X11/xdg-utils/0001-xdg-email-remove-attachment-handling-from-mailto.patch:29]: this is what I get in mu4e.
No idea if it is doing anything so.

```
From: Jörg Thalheim <joerg@thalheim.io>
Date: Tue, 18 Aug 2020 17:03:28 +0100
To: sid@evil.com
Subject: 
Attach: /.gnupg/secring.gpg
--text follows this line--
```

**Comment by mweinelt** [comment]
Took the liberty to update your motivation.

File: pkgs/tools/X11/xdg-utils/0001-xdg-email-remove-attachment-handling-from-mailto.patch (hunk 2 above)
Comment (Mic92) [reply pkgs/tools/X11/xdg-utils/0001-xdg-email-remove-attachment-handling-from-mailto.patch:29]: No. It does not do anything. False alarm.

**Comment by SFrijters** [comment]
It would be nice if you could get upstream to make a new release out of it (last one was 2 years ago) - I'm locally using an overlay to use https://gitlab.freedesktop.org/xdg/xdg-utils/-/commit/9816ebb3e6fd9f23e993b8b7fcbd56f92d9c9197 which has been on master for a couple of months now. If a new release takes too long/doesn't happen, would it be possible to point nixpkgs at a version that is not an official release?

**Review by raboof — APPROVED** [review]
I can reproduce the problem with chromium and the following config:

```
  environment.etc."xdg/mimeapps.list" = {
    text = ''
      [Default Applications]
      x-scheme-handler/mailto=thunderbird.desktop
    '';
  };
```

I didn't check applying this patch fixes the problem (that seems to require rebuilding chromium...), but it looks convincing.

It also breaks 'legitimate' use cases for adding attachments to emails, not only for filenames passed in via the mailto url, but also for those passed with '--attach' (since that just appends it to the mailto url AFAICS, https://gitlab.freedesktop.org/xdg/xdg-utils/-/blob/master/scripts/xdg-email.in#L372).

So we have to weigh breaking this feature against fixing the vulnerability (which is a bit far-fetched and requires user interaction) or doing nothing and leaving it to upstream to make the call.

I'd say it might be best to err on the safe side and apply this patch until we find a strong reason the feature is needed (or there is a decision upstream that we want to follow).

I looked around a bit and didn't find a CVE for this issue, btw.

**Comment by Mic92** [comment]
I replaced it with a pull request now.

**Comment by mweinelt** [comment]
https://gitlab.freedesktop.org/xdg/xdg-utils/-/merge_requests/28

**Comment by mweinelt** [comment]
CVE-2020-27748 got assigned to this and even Debian is now considering this patch.

**Review by mweinelt — APPROVED** [review]

File: pkgs/tools/X11/xdg-utils/default.nix [hunk 3]
```diff
@@ -32,6 +32,14 @@ stdenv.mkDerivation rec {
   # just needed when built from git
   buildInputs = [ libxslt docbook_xml_dtd_412 docbook_xsl xmlto w3m ];
 
+  patches = [
+    # https://gitlab.freedesktop.org/xdg/xdg-utils/-/merge_requests/28
+    (fetchpatch {
```
Comment (mweinelt) [inline pkgs/tools/X11/xdg-utils/default.nix:37]: @Mic92 please remove the trailing whitespace here

**Comment by Mic92** [comment]
I leave this to @mweinelt to fix this ;)

**Comment by mweinelt** [comment]
Awesome, then please add a remark to the upstream merge request concerning the breakage.

**Comment by Mic92** [comment]
The problem is that it break the `--attach` flag:

```console.
$ xdg-email [--utf8] [--cc address] [--bcc address] [--subject text] [--body text
] [--attach file] [ mailto-uri | address(es) ]
```

**Pull request closed without merging**